src/components/checkbox/GroupableCheckboxReducers.ts

```diff
@@ -0,0 +1,103 @@
+
+export interface IGroupableCheckboxesState {
+  total: number;
+  nbChecked: number;
+  parentId?: string;
+  parent?: ICheckboxState;
+  checkboxes: ICheckboxState[];
+}
+
+export const groupableCheckboxInitialState: IGroupableCheckboxesState = {total: 0, nbChecked: 0, checkboxes: []};
+export const groupableCheckboxesInitialState: IGroupableCheckboxesState[] = [];
+
+export const groupableCheckboxReducer = (state: IGroupableCheckboxesState = groupableCheckboxInitialState,
+                                         action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState => {
+  switch (action.type) {
+    case GroupableCheckboxActions.addGroup:
+      if (action.payload.isParent) {
+        return addParentCheckbox(state, action);
+      }
+      return addChildCheckbox(state, action);
+
+    case GroupableCheckboxActions.toggleGroup:
+      if (state.parentId === action.payload.id || state.parentId === action.payload.parentId) {
+        let isChecked: boolean;
+        if (action.payload.isParent) {
+          isChecked = !state.parent.checked;
+          return {
+            ...state,
+            parent: {
+              id: action.payload.id,
+              checked: isChecked,
+            },
+            checkboxes: _.map(state.checkboxes, (checkbox: ICheckboxState) => {
+              checkbox.checked = isChecked;
+              return checkbox;
+            }),
+            nbChecked: isChecked ? state.total : 0,
+          };
+        }
+        const checkboxes: ICheckboxState[] = _.map(state.checkboxes, (checkbox: ICheckboxState) => {
+          if (checkbox.id === action.payload.id) {
+            isChecked = !checkbox.checked;
+            return {
+              id: checkbox.id,
+              checked: isChecked,
+            };
+          }
+          return checkbox;
+        });
+        if (checkboxes) {
+          const nbChecked: number = state.nbChecked += (isChecked ? 1 : -1);
+          return {
+            ...state,
+            parent: {
+              id: action.payload.parentId,
+              checked: nbChecked === state.total,
+            },
+            nbChecked,
+            checkboxes: [...checkboxes],
+          };
+        }
+      }
+      return state;
+    default:
+      return state;
+  }
+};
+
+export const groupableCheckboxesReducer = (state: IGroupableCheckboxesState[] = groupableCheckboxesInitialState,
+                                           action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState[] => {
+  let groupableCheckboxesState: IGroupableCheckboxesState;
+  switch (action.type) {
+    case GroupableCheckboxActions.addGroup:
+      groupableCheckboxesState = _.findWhere(state, { parentId: (action.payload.parentId || action.payload.id) });
+      if (state && groupableCheckboxesState) {
+        groupableCheckboxReducer(groupableCheckboxesState, action);
+        return state;
+      }
```

Comment:
Why are you returning the state here?

Reply:
because we only update the state because its already existe by a previous checkbox ( child or parent ) added. I only need to update the state with the new one.

Reply:
You aren't really updating the state if you are returning the old state.